cli/best_ancestor.go

```diff
@@ -118,34 +103,92 @@ func fastFindAncestor(r *git.Repository) (h plumbing.Hash, c int, branchName str
 	return
 }
 
-func findRemoteBranch(reference *plumbing.Reference, r *git.Repository) (remoteBranch *plumbing.Reference, err error) {
+func (p *RemoteCmd) findRemoteBranch(ctx context.Context, reference *plumbing.Reference, r *git.Repository) (remoteBranch *plumbing.Reference, err error) {
 	branchIter, err := r.References()
 	if err != nil {
 		return
 	}
 	// If branch is main, just point to it directly
-	remoteDefault := "refs/remote/origin/main"
+	remoteDefault := "refs/remotes/origin/main"
 	remoteBranch, err = r.Reference(plumbing.ReferenceName(remoteDefault), false)
 	if err == nil {
 		return
 	}
 	// If branch is master, just point to it directly
-	remoteDefault = "refs/remote/origin/master"
+	remoteDefault = "refs/remotes/origin/master"
 	remoteBranch, err = r.Reference(plumbing.ReferenceName(remoteDefault), false)
 	if err == nil {
 		return
 	}
 	// We have some issues here
-	// TODO(ch2036): Maybe switch to using `git` utility directly
-	err = branchIter.ForEach(func(rem *plumbing.Reference) error {
-		if rem.Name().IsRemote() {
-			log.Debugf("Branch found: %v, %s ?== %s", rem, rem.Name().Short(), reference.Name().Short())
-			if strings.HasSuffix(rem.Name().Short(), reference.Name().Short()) {
-				remoteBranch = rem
-				return storer.ErrStop
+	if p.goGit {
+		log.Warnf("go-git traverse of references can lead to issues...")
+		err = branchIter.ForEach(func(rem *plumbing.Reference) error {
+			if rem.Name().IsRemote() {
+				log.Debugf("Branch found: %v, %s ?== %s", rem, rem.Name().Short(), reference.Name().Short())
+				if strings.HasSuffix(rem.Name().Short(), reference.Name().Short()) {
+					remoteBranch = rem
+					return storer.ErrStop
+				}
+			}
+			return nil
+		})
+	} else {
+		const (
+			gitFetchCmd  = " fetch --all"
```

Comment:
If you're just looking to list the remote branches, you can use [`git ls-remote`](https://git-scm.com/docs/git-ls-remote.html) without performing a fetch.

Reply:
We need to perform a fetch, because we don't know for sure if the local clone was a shallow one or not. (But, if he/she created a new topic branch, most probably the local repo, at `$(pwd)/.git` has enough info to find it... got your point, I didn't thought so previously). This could simplify remote build logic in other places too.